### PR TITLE
core: add sender field to transaction

### LINKED ISSUE #860: Add transaction Sender field
In NEO 3 every transaction has an explicit `Sender` that signs it, thus we need to add this field and add verification of `Sender`'s signature. Relates to #831.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #860

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
This one could also be improved by adding SendersFeeMonitor (like in https://github.com/neo-project/neo/commit/f1d64343) in order not to re-calculate values of `netFee` and `sysFee` for same account each time new transaction is being verified. Should I also add SendersFeeMonitor?

File: pkg/core/transaction/transaction.go [hunk 1]
```diff
@@ -118,6 +121,16 @@ func (t *Transaction) DecodeBinary(br *io.BinReader) {
 	t.Type = TXType(br.ReadB())
 	t.Version = uint8(br.ReadB())
 	t.Nonce = br.ReadU32LE()
+
+	bytes := make([]byte, util.Uint160Size)
+	br.ReadBytes(bytes)
+	sender, err := util.Uint160DecodeBytesLE(bytes)
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go:127]: 1) It should be BE
2) Take a look at `output.go` and the way `ScriptHash` is being handled there, way simpler than this code.

File: pkg/rpc/request/txBuilder.go [hunk 2]
```diff
@@ -39,6 +39,7 @@ func CreateRawContractTransaction(params ContractTxParams) (*transaction.Transac
 		return nil, errs.Wrapf(err, "Failed to take script hash from address: %v", toAddress)
 	}
 	tx.AddVerificationHash(fromAddressHash)
```
Comment (roman-khimov) [inline pkg/rpc/request/txBuilder.go:41]: Shouldn't be needed now.

File: pkg/rpc/client/rpc.go [hunk 3]
```diff
@@ -517,17 +517,19 @@ func (c *Client) SignAndPushInvocationTx(script []byte, acc *wallet.Account, sys
...
 	if gas > 0 {
 		if err = request.AddInputsAndUnspentsToTx(tx, acc.Address, core.UtilityTokenID(), gas, c); err != nil {
 			return txHash, errors.Wrap(err, "failed to add inputs and unspents to transaction")
 		}
 	} else {
-		addr, err := address.StringToUint160(acc.Address)
-		if err != nil {
-			return txHash, errors.Wrap(err, "failed to get address")
-		}
 		tx.AddVerificationHash(addr)
```
Comment (roman-khimov) [inline pkg/rpc/client/rpc.go:533]: Should be dropped too.

File: pkg/core/transaction/state.go [hunk 4]
```diff
@@ -17,6 +18,7 @@ func NewStateTX(state *StateTX) *Transaction {
 		Type:       StateType,
 		Version:    0,
 		Nonce:      rand.Uint32(),
+		Sender:     util.Uint160{},
```
Comment (roman-khimov) [inline pkg/core/transaction/state.go:21]: I don't think it makes sense adding these, that's the default value anyway.

File: pkg/core/blockchain.go [hunk 5]
```diff
@@ -1316,6 +1317,25 @@ func (bc *Blockchain) verifyTx(t *transaction.Transaction, block *block.Block) e
...
+		return errors.Errorf("insufficient funds: balanse is %v, fee is %v", balance, fee)
+	}
+	for _, tx := range bc.memPool.GetVerifiedTransactions() {
+		if tx.Tx.Sender == t.Sender && tx.Tx != t {
+			fee += tx.Fee + bc.SystemFee(tx.Tx)
+		}
+	}
+	if balance < fee {
+		return errors.Errorf("insufficient funds: balanse is %v, total sender fee is %v", balance, fee)
+	}
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1337]: I think we can completely drop this part until we have sys fee and net fee fields in `Transaction` (#831).

**Comment by roman-khimov** [comment]
> This one could also be improved by adding SendersFeeMonitor

Not yet, it also is for #831, but even there it should be done in a bit different manner.

File: pkg/core/blockchain.go [hunk 6]
```diff
@@ -1868,6 +1868,11 @@ func (bc *Blockchain) GetScriptHashesForVerifying(t *transaction.Transaction) ([
 			}
 		}
 	}
+
+	sender := t.Sender
+	if _, ok := hashes[sender]; !ok {
+		hashes[sender] = true
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1874]: Why not just `hashes[t.Sender] = true`?

File: pkg/consensus/consensus_test.go [hunk 7]
```diff
@@ -259,3 +266,38 @@ func (fs *feer) NetworkFee(*transaction.Transaction) util.Fixed8 { return util.F
...
+func addSender(t *testing.T, txs ...*transaction.Transaction) {
+	// multisig address which possess all NEO
+	scriptHash, err := util.Uint160DecodeStringBE("d60ac443bb800fb08261e75fa5925d747d485861")
+	require.NoError(t, err)
+	for _, tx := range txs {
+		tx.Sender = scriptHash
+	}
+}
+
+func signTx(t *testing.T, txs ...*transaction.Transaction) {
```
Comment (roman-khimov) [inline pkg/consensus/consensus_test.go:279]: Technically it'd be more appropriate to add signature of one of CNs, not multisig, but this code will go away soon (with Miner tx itself), so probably it's not worth fixing it now.

File: pkg/core/blockchain.go [hunk 8]
```diff
@@ -1927,7 +1932,6 @@ func (bc *Blockchain) GetScriptHashesForVerifying(t *transaction.Transaction) ([
 	}
 
 	return hashesResult, nil
-
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1930]: Useless change.

**Pull request merged**